lib/datadog/core/configuration/option.rb

```diff
@@ -94,6 +96,93 @@ def default_precedence?
 
         private
 
+        def coerce_env_variable(value)
+          return context_exec(value, &@definition.env_parser) if @definition.env_parser
+
+          case @definition.type
+          when :int
+            # DEV-2.0: Change to a more strict coercion method. Integer(value).
+            value.to_i
+          when :float
+            # DEV-2.0: Change to a more strict coercion method. Float(value).
+            value.to_f
+          when :array
+            values = if value.include?(',')
+                       value.split(',')
+                     else
+                       value.split(' ') # rubocop:disable Style/RedundantArgument
+                     end
+
+            values.map! do |v|
+              v.gsub!(/\A[\s,]*|[\s,]*\Z/, '')
+
+              v.empty? ? nil : v
+            end
+
+            values.compact!
+            values
+          when :bool
+            string_value = value
+            string_value = string_value.downcase
+            string_value == 'true' || string_value == '1' # rubocop:disable Style/MultipleComparison
+          when :string, NilClass
+            value
+          else
+            raise ArgumentError, "The option #{@definition.name} is using an unknown type option `#{@definition.type}`"
+          end
+        end
+
+        def validate_type(value)
+          raise_error = false
+
+          valid_type = validate(@definition.type, value)
+
+          unless valid_type
+            raise_error = if @definition.type_options[:nilable]
+                            !value.is_a?(NilClass)
+                          else
+                            true
+                          end
+          end
+
+          if raise_error
+            error_msg = if @definition.type_options[:nilable]
+                          "The option #{@definition.name} support this type `#{@definition.type}` "\
+                                      "and `nil` but the value provided is #{value.class}"
+                        else
+                          "The option #{@definition.name} support this type `#{@definition.type}` "\
+                          "but the value provided is #{value.class}"
+                        end
```

Comment:
Right now this is not particularly user-friendly:

```
$ cat example.rb 
require 'ddtrace'

Datadog.configure { |c| c.profiling.enabled = 'potato' }
$ bundle exec ruby example.rb 
Traceback (most recent call last):
  7: from example.rb:3:in `<main>'
  6: from lib/datadog/tracing/contrib/extensions.rb:64:in `configure'
  5: from lib/datadog/core/configuration.rb:82:in `configure'
  4: from example.rb:3:in `block in <main>'
  3: from lib/datadog/core/configuration/options.rb:47:in `block in default_helpers'
  2: from lib/datadog/core/configuration/options.rb:70:in `set_option'
  1: from lib/datadog/core/configuration/option.rb:54:in `set'
lib/datadog/core/configuration/option.rb:157:in `validate_type': The option enabled support this type `bool` but the value provided is String (ArgumentError)
```

I suggest approaching this from the POV of a user getting this error and what they'd need to fix it:

```
The setting `.profiling.enabled` inside the `Datadog.configure` block expects a bool, but a `String` was provided ('potato'). Please update your `configure` block. Alternatively, you can disable this validation using `...env var here...`. For help, see <link to section of doc where we can tell users to get in touch with support>.
```

I did a quick experiment and I could inject the name of the containing group (although not the full path) by doing

```diff
           # e.g. `settings :foo { option :bar }` --> `config.foo.bar`
           # @param [Symbol] name option name. Methods will be created based on this name.
           def settings(name, &block)
-            settings_class = new_settings_class(&block)
+            settings_class = new_settings_class(name, &block)
 
             option(name) do |o|
               o.default { settings_class.new }
@@ -39,8 +39,9 @@ module Datadog
 
           private
 
-          def new_settings_class(&block)
+          def new_settings_class(name = nil, &block)
             Class.new { include Configuration::Base }.tap do |klass|
+              klass.instance_variable_set(:@name, name)
               klass.instance_eval(&block) if block
```

which is why I was able to prototype `.profiling.enabled`.

Reply:
Done on https://github.com/DataDog/dd-trace-rb/pull/2983